core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/scheduling/CostBasedScheduleGenerator.scala

```diff
@@ -83,16 +88,33 @@ class CostBasedScheduleGenerator(
           })
           .filter(link => operatorIds.contains(link.fromOpId))
         val operators = operatorIds.map(operatorId => physicalPlan.getOperator(operatorId))
-        val materializedPortIds: Set[GlobalPortIdentity] = matEdges.flatMap(link =>
-          List(
-            GlobalPortIdentity(link.fromOpId, link.fromPortId, input = false)
+        val portIdsToViewResult: Set[GlobalPortIdentity] =
+          workflowContext.workflowSettings.outputPortsToViewResult
+            .filter(outputPort => operatorIds.contains(outputPort.opId))
+        val portIdsNeedingStorage: Set[GlobalPortIdentity] = matEdges
+          .diff(physicalPlan.getDependeeLinks)
+          .filter(matLink => operatorIds.contains(matLink.fromOpId))
+          .flatMap(link =>
+            List(
+              GlobalPortIdentity(link.fromOpId, link.fromPortId)
+            )
+          ) ++ portIdsToViewResult
+        val outputPortResultURIs = portIdsNeedingStorage
+          .map(outputPortId =>
+            outputPortId -> createResultURI(
+              workflowId = workflowContext.workflowId,
+              executionId = workflowContext.executionId,
+              operatorId = outputPortId.opId.logicalOpId,
+              layerName = Some(outputPortId.opId.layerName),
+              portIdentity = outputPortId.portId
+            )
           )
-        )
+          .toMap
         Region(
           id = RegionIdentity(idx),
           physicalOps = operators,
           physicalLinks = links,
-          materializedPortIds = materializedPortIds
+          outputPortResultURIs = outputPortResultURIs
```

Comment:
is the URI object serializable? we may need to serialize a Region. if not, consider using string instead.

Reply:
also, just call it storageURIs? it is of course related to `outputPort`, and not necessarily related to `result` (as it could be used for internal materializations). In general, I suggest in the engine, we don't call anything `result` as this is a relative term, those materializations could be input to some downstream, rather than just the result of some upstream.

Reply:
URI object is serializable as it is a `java.io.Serializable`. Changed name to `storageURIs `.